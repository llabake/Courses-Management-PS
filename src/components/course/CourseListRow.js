/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const CourseListRow = ({ course, onDelete }) => {
  const deleteCourse = () => (
    onDelete(course.id)
  );
  return (
    <tr>
      <td><a href={`/course/${course.watchHref}`} target="_blank" rel="noopener noreferrer">Watch</a></td>
      <td><Link to={`/course/${course.id}`}>{course.title}</Link></td>
      <td>{course.authorId}</td>
      <td>{course.category}</td>
      <td>{course.length}</td>
      <td
        className="btn btn-xs btn-danger"
        onClick={deleteCourse}
      >
Delete
      </td>
    </tr>
  );
};

CourseListRow.propTypes = {
  course: PropTypes.object.isRequired,
  onDelete: PropTypes.func.isRequired,
};

export default CourseListRow;
