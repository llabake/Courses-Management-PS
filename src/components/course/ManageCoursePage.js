import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Prompt } from 'react-router-dom';
import toastr from "toastr";
import {
  bool,
  shape,
  arrayOf,
  func,
} from 'prop-types';
import { saveCourse } from '../../actions/courseActions';
import CourseForm from "./CourseForm";

export class ManageCoursePage extends Component {
  state = {
    course: Object.assign({}, this.props.course),
    errors: {},
    saving: false,
    isBlocking: false
  };

  // eslint-disable-next-line consistent-return
  componentWillReceiveProps = (nextProps) => {
    if (this.props.course.id !== nextProps.course.id) {
      return this.setState({
        course: Object.assign({}, nextProps.course),
      });
    }
  };

  updateCourseState = (event) => {
    const field = event.target.name;
    const { course } = this.state;
    course[field] = event.target.value;
    return this.setState({ course, isBlocking: true, });
  };

  redirect() {
    this.setState({
      saving: false,
      isBlocking: false,
    });
    toastr.success("Course saved successfully");
    this.props.history.push('/courses');
  }

  saveCourse = (event) => {
    event.preventDefault();
    this.setState({ saving: true });
    this.props.saveCourse(this.state.course)
      .then(() => this.redirect())
      .catch((error) => {
        toastr.error(error);
        this.setState({ saving: false });
      });
  };

  render() {
    const { course, errors, isBlocking } = this.state;
    const { authors } = this.props;
    return (
      <div>
        <Prompt
          when={isBlocking}
          message={"Are you sure you want to go to leave the page"}
        />
        <CourseForm
          course={course}
          errors={errors}
          allAuthors={authors}
          onChange={this.updateCourseState}
          onSave={this.saveCourse}
          saving={this.state.saving}
        />
      </div>
    );
  }
}

const getCourseById = (courses, id) => {
  const course = courses.filter(element => element.id === id);
  if (course.length) return course[0];
  return null;
};


export const mapStateToProps = (state, { match: { params } }) => {
  const courseId = params.id;
  let course = {
    id: '', watchHref: '', title: '', authorId: '', length: '', category: '',
  };

  if (courseId && state.courses.length > 0) {
    course = getCourseById(state.courses, courseId);
  }

  const authorsFormattedForDropdown = state.authors.map(author => ({
    value: author.id,
    text: `${author.firstName} ${author.lastName}`,
  }));

  return {
    course,
    authors: authorsFormattedForDropdown,
  };
};

export const mapDispatchToProps = dispatch => ({
  saveCourse: course => dispatch(saveCourse(course)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ManageCoursePage);
