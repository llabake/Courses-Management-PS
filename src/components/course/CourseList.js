import React from 'react';
import PropTypes from 'prop-types';
import CourseListRow from "./CourseListRow";

const CourseList = ({ courses, onDelete }) => (
  <table className="table">
    <thead>
      <tr>
        <th>&nbsp;</th>
        <th>Title</th>
        <th>Author</th>
        <th>Category</th>
        <th>Length</th>
        <th>Delete</th>
      </tr>
    </thead>
    <tbody>
      {courses.map(course => (
        <CourseListRow
        key={course.id}
        course={course}
        onDelete={onDelete} />
      ))}
    </tbody>
  </table>
);

CourseList.propTypes = {
  courses: PropTypes.array.isRequired,
};

export default CourseList;
