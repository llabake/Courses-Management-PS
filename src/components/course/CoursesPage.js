import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import toastr from "toastr";
import { deleteCourse } from '../../actions/courseActions';
import CourseList from './CourseList';
import { sortCourses } from "../../helpers";
import NoCourses from "../common/NoCourses";

class CoursesPage extends Component {
  static courseRow(course, index) {
    return <div key={index}>{course.title}</div>;
  }

  state = {
    course: {
      title: "",
    },
  };

  redirectToAddCoursePage = () => {
    const { history } = this.props;
    history.push('/course');
  };

  deleteCourse = (courseId) => {
    const { history } = this.props;
    this.props.deleteCourse(courseId).then(() => {
      toastr.success('Course deleted successfully');
      history.push('/courses');
    });
  };

  render() {
    const { courses, loading } = this.props;
    return (
      <div>
        <h1>Courses</h1>
        <input
          type="submit"
          value="Add Course"
          className="btn btn-primary"
          onClick={this.redirectToAddCoursePage}
        />
        { !loading && !courses.length
          ? <NoCourses
              text={'There are no courses now. Click the Add Course button to add courses'} />
          : <CourseList courses={courses} onDelete={this.deleteCourse} />}
      </div>
    );
  }
}

export const mapStateToProps = (state) => {
  const { courses } = state;
  sortCourses(courses);
  return {
    courses,
    loading: state.ajaxCallsInProgress > 0,
  };
};


export const mapDispatchToProps = dispatch => ({
  deleteCourse: course => dispatch(deleteCourse(course)),
});

export default connect(mapStateToProps, mapDispatchToProps)(CoursesPage);
