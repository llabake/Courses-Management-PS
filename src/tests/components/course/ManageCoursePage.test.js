import React from 'react';
import { shallow } from 'enzyme';
import { ManageCoursePage, mapStateToProps, mapDispatchToProps } from "../../../components/course/ManageCoursePage";
import courseMock from "../../mocks/courseMock";
import authorMock from "../../mocks/authorMock";

describe('<ManageCoursePage', () => {
  it('renders correctly', () => {
    const state = {
      course: {},
      errors: {},
      saving: false,
      isBlocking: false,
      isValid: false,
    };
    const wrapper = shallow(<ManageCoursePage {...state} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('calls the componentWillReceiveProps method', () => {
    const state = {
      course: {},
      errors: {},
      saving: false,
      isBlocking: false,
      isValid: false,
    };
    const wrapper = shallow(<ManageCoursePage {...state} />);
    const spy = jest.spyOn(wrapper.instance(), 'componentWillReceiveProps');
    const incomingProps = {
      course: courseMock.courses[0],
    };
    wrapper.instance().componentWillReceiveProps(incomingProps);
    expect(spy).toHaveBeenCalled();
  });
  it('should save a course', () => {
    const props = {
      course: {},
      match: {
        params: {
          id: courseMock.courses[0].id,
        },
      },
      saveCourse: jest.fn(() => Promise.resolve()),
      updateCourseState: jest.fn(),
      history: { push: jest.fn() },
      getCourseById: jest.fn((courses, id) => {
        const course = courses.filter(element => element.id === id);
        if (course.length) return Promise.resolve(course[0]);
        return Promise.reject();
      }),
    };
    const wrapper = shallow(<ManageCoursePage {...props} />);
    const spy = jest.spyOn(wrapper.instance(), 'saveCourse');
    wrapper.setState({
      course: {
        title: 'my course title',
        length: '07:10',
        authorId: 'my author',
        category: 'my category',
      },
    });
    wrapper.instance().saveCourse({ preventDefault: jest.fn() });
    expect(spy).toHaveBeenCalled();
  });
  it('renders a Prompt child component', () => {
    const state = {
      course: {},
      errors: {},
      saving: false,
      isBlocking: false,
      isValid: false,
    };

    const props = {
      saveCourse: jest.fn(),
      updateCourseState: jest.fn(),
      history: { push: jest.fn() },
      getCourseById: jest.fn((courses, id) => {
        const course = courses.filter(element => element.id === id);
        if (course.length) return Promise.resolve(course[0]);
        return Promise.reject();
      }),
    };
    const wrapper = shallow(<ManageCoursePage {...state} {...props} />);
    expect(wrapper.find('Prompt').length).toBe(1);
  });
  it('should update a course', () => {
    const props = {
      course: courseMock.courses[0],
      authors: authorMock.authors,
      match: {
        params: {
          id: courseMock.courses[0].id,
        },
      },
      saveCourse: jest.fn(() => Promise.resolve()),
      updateCourseState: jest.fn(() => Promise.resolve()),
      history: { push: jest.fn() },
      getCourseById: jest.fn((courses, id) => {
        const course = courses.filter(element => element.id === id);
        if (course.length) return Promise.resolve(course[0]);
        return Promise.reject();
      }),
    };
    const wrapper = shallow(<ManageCoursePage {...props} />);
    const spy = jest.spyOn(wrapper.instance(), 'updateCourseState');
    const event = {
      preventDefault: jest.fn(),
      target: {
        name: 'title',
        value: 'this girl',
      },
    };
    wrapper.setState({
      course: {
        title: 'my course title',
        length: '07:10',
        authorId: 'my author',
        category: 'my category',
      },
    });
    wrapper.instance().updateCourseState(event);
    expect(spy).toHaveBeenCalled();
  });
  it('should dispatch needed actions', () => {
    const dispatch = jest.fn();
    expect(mapDispatchToProps(dispatch)).toHaveProperty('saveCourse');
    const { saveCourse } = mapDispatchToProps(dispatch);
    saveCourse(courseMock.createdCourseWithId);
    expect(dispatch).toHaveBeenCalled();
  });
});
