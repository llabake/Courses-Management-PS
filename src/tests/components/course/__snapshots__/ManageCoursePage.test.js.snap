// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ManageCoursePage renders correctly 1`] = `
<div>
  <Prompt
    message="Are you sure you want to go to leave the page"
    when={false}
  />
  <CourseForm
    course={Object {}}
    errors={Object {}}
    onChange={[Function]}
    onSave={[Function]}
    saving={false}
  />
</div>
`;
